### DISCUSSION TITLE
Firebase Functions 2nd Gen: The request was not authorized to invoke this service.

### DISCUSSION THREAD

**Discussion by madhavgharmalkar** [firebase/firebase-ios-sdk#12305] [Q&A]
Hey! 👋🏾 

I'm running into an issue using Swift, and 2nd Gen firebase functions (Node / Typescript). I've successfully tested my functions using the local emulator, and have successfully deployed them to firebase. However when I try to invoke the function, I run into the following error:

```The request was not authorized to invoke this service. Read more at https://cloud.google.com/run/docs/securing/authenticating Additional troubleshooting documentation can be found at: https://cloud.google.com/run/docs/troubleshooting#401```

After doing some searching on google, I found [this StackOverflow issue](https://stackoverflow.com/questions/76380806/firebase-cloud-functions-v2-the-request-was-not-authorized-to-invoke-this-servi) which recommends adding the `Cloud Functions Invoker` permission to the service account. I tried that and it didn't work. Just to make sure, I added the `Cloud Functions Invoker` role to every account in my GCP IAM and I still haven't had any luck.

<img width="530" alt="image" src="https://github.com/firebase/firebase-ios-sdk/assets/8007933/44208783-2fd3-42d8-949d-106aa2f905a2">

Any suggestions would be awesome!

**Comment by elishaJ** [comment]
@madhavgharmalkar 👋🏽 I came here trying to find a solution myself. I was getting the same error when deploying 2nd gen Cloud function (written in Python). I managed to resolve it by updating the IAM policy. I had to update the members from service account to allUsers (since I built a public API) and assigning the roles/run.invoker. 

![image](https://github.com/firebase/firebase-ios-sdk/assets/28643263/8335d7a5-ce7a-4a44-a5dc-abe962e83195)


Verify the policy using:

`gcloud run services get-iam-policy projects/<project-id>/locations/<region>/services/<function-name>
bindings:
- members:
  - allUsers
  role: roles/run.invoker`
  
 Try assigning this role to the user/service account according to your function requirements. I hope that helps 🙌🏽

**Comment by dahed92** [comment]
What worked for me was deleting my functions from the google cloud console and redeploy them

**Reply by guihuaz** [reply]
hi @dahed92 ,  Could you please share what you did before deleting functions and redeploy them? Did you add `Cloud Functions Invoker` permission to the service account? I have same 401 issue and still couldn't figure out.

**Reply by dahed92** [reply]
Hi, it was my personal project, so I had all the necessary permissions, and I believe I added the Invoker permission. In any case, try this process both with and without the permission: Go to GCP, search for "Cloud Run," and you'll see a list of all your cloud functions. Select them all and delete them. Make sure you have a local copy of your functions before doing this.

Hope this helps!
